### PR TITLE
HPCC-24896 Base64 encode hashed user password in JWT SecMgr plugin

### PR DESCRIPTION
**Pull request by dcamper**
Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by dcamper** [comment]
@ghalliday Please review.  Thank you!

**Comment by richardkchapman** [comment]
This should probably target 7.12.x ?

**Comment by dcamper** [comment]
Argh.  You're right, Richard.  Rebased and retargeted to 7.12.x.

File: system/security/plugins/jwtSecurity/jwtEndpoint.cpp [hunk 1]
```diff
@@ -50,28 +50,29 @@ static size_t captureIncomingCURLReply(void* contents, size_t size, size_t nmemb
...
 
-    return hashedValue;
+    return std::string(hashedValue, SHA256_DIGEST_LENGTH);
 }
 
-static std::string hashUserPW(const std::string& pw, const std::string& nonce)
+static void hashUserPW(const std::string& pw, const std::string& nonce, StringBuffer& encodedHash)
 {
-    std::string     firstHash(hashString(pw));
+    std::string     saltedPW(pw.substr(0, 2) + pw);
```
Comment (ghalliday) [inline system/security/plugins/jwtSecurity/jwtEndpoint.cpp:71]: This doesn't seem to be directly related, and it doesn't seem to be a conventional password SALT.  Is it correct?

**Review by ghalliday — COMMENTED** [review]
@dcamper the change looks sensible.  One question.

File: system/security/plugins/jwtSecurity/jwtEndpoint.cpp (hunk 1 above)
Comment (dcamper) [reply system/security/plugins/jwtSecurity/jwtEndpoint.cpp:71]: I should have noted that change in the Jira.  The reason behind it is, of course, to mitigate rainbow table attacks, both for transmitted and stored passwords.  Random values are preferred for salts but cannot be used in this unusual configuration, where the client is capturing credentials but authenticating happens remotely (well, perhaps you could, but it would involve an extra message exchange for every auth to trade random values).  In a full OpenID Connect implementation the user would present credentials only at the auth server, where random salts could be used.

This type of salt does have a security flaw in that an attacker with access to the saved passwords will be able to see if multiple users have the same password, even if they cannot use (common) rainbow tables to look up the plaintext version.

TL;DR: A two-character salt based on the user's pw value is not the best but it is better than an unsalted value and the compromise fits into our modified OpenID Connect auth scheme.

If you have a better idea for the above, let me know!

Comment (ghalliday) [reply system/security/plugins/jwtSecurity/jwtEndpoint.cpp:71]: It sounds like a decent compromise.  It might be worth adding a comment to explain.  I can merge with or without the comment.

Comment (dcamper) [reply system/security/plugins/jwtSecurity/jwtEndpoint.cpp:76]: Comment added and squashed.
